gunrock/cuda/launch_box.hxx

```diff
@@ -12,44 +12,148 @@
+#ifdef __CUDA_ARCH__
+#if   __CUDA_ARCH__ >= 860
+  #define SM_TAG   sm_86
+  #define ARCH_TAG ampere
+#elif __CUDA_ARCH__ >= 800
+  #define SM_TAG   sm_80
+  #define ARCH_TAG ampere
+#elif __CUDA_ARCH__ == 750
+  #define SM_TAG   sm_75
+  #define ARCH_TAG turing
+#elif __CUDA_ARCH__ >= 700
+  #define SM_TAG   sm_70
+  #define ARCH_TAG volta
+#elif __CUDA_ARCH__ == 620
+  #define SM_TAG   sm_62
+  #define ARCH_TAG pascal
+#elif __CUDA_ARCH__ >= 610
+  #define SM_TAG   sm_61
+  #define ARCH_TAG pascal
+#elif __CUDA_ARCH__ >= 600
+  #define SM_TAG   sm_60
+  #define ARCH_TAG pascal
+#elif __CUDA_ARCH__ == 530
+  #define SM_TAG   sm_53
+  #define ARCH_TAG maxwell
+#elif __CUDA_ARCH__ >= 520
+  #define SM_TAG   sm_52
+  #define ARCH_TAG maxwell
+#elif __CUDA_ARCH__ >= 500
+  #define SM_TAG   sm_50
+  #define ARCH_TAG maxwell
+#elif __CUDA_ARCH__ == 370
+  #define SM_TAG   sm_37
+  #define ARCH_TAG kepler
+#elif __CUDA_ARCH__ >= 350
+  #define SM_TAG   sm_35
+  #define ARCH_TAG kepler
+#elif __CUDA_ARCH__ == 320
+  #define SM_TAG   sm_32
+  #define ARCH_TAG kepler
+#elif __CUDA_ARCH__ >= 300
+  #define SM_TAG   sm_30
+  #define ARCH_TAG kepler
+#else
+  #error "Gunrock only supports sm_30 and above"
+#endif  // __CUDA_ARCH__
+#else
+  // What should these macros be on the host side?
+  #define SM_TAG   sm_00
+  #define ARCH_TAG ampere
+#endif
+
 namespace gunrock {
 namespace cuda {
 
-// Should the struct members be template parameters?
-template<int blockDim_, int gridDim_, size_t smemBytes_ = 0>
-struct launch_params_t{
-  int blockDim = blockDim_;
-  int gridDim = gridDim_;
-  size_t smemBytes = smemBytes_;
-};
+//////////////// Move this section to another file?
+/**
+ * @brief Blank struct to use as a default base to make inheritance optional
+ */
+struct empty_t {};
 
-// template<int ccCombined, int... launch_tparams>
-// struct cc_launch_params_t : launch_params_t<launch_tparams...> {
-//     // Compute capability related things
-// };
+// Inheritance structs to expand the variadic launch types passed into the launch box and define each them within the launch box's namespace
+template<typename... base_v>
+struct inherit_t;
 
-// smemBytes_ isn't an int, but I am assuming it will get cast to a size_t? I think there is a better way to generalize the type
-template<int... launch_tparams>
-struct arch_launch_params_t : launch_params_t<launch_tparams...> {
-  // Architecture related things
+template<typename base_t, typename... base_v>
+struct inherit_t<base_t, base_v...> :
+base_t::template rebind<inherit_t<base_v...>> {};  // What does ::template do here?
+
+template<typename base_t>
+struct inherit_t<base_t> : base_t {};
+////////////////
+
+/**
+ * @brief Struct holding kernel parameters will be passed in upon launch
+ * @tparam blockDim_ 1D block dimensions to launch with
+ * @tparam gridDim_ 1D grid dimensions to launch with
+ * @tparam smemBytes_ Amount of shared memory to allocate
+ */
+template<
+  unsigned int blockDim_,
```

Comment:
Let's stay consistent with naming convention; `block_dimension` or `number_of_blocks` and not `blockDim` or `testTest`.

Reply:
Done